### PR TITLE
Evict content indirectly referenced via symlinks from VFS

### PR DESCRIPTION
**Pull request by lptr**
Before this change we evicted content from the VFS that was referenced directly via symlinks to avoid relying on cached file system data that might change without notification. However, content that is marked as directly accessed in the VFS might still be referenced via symlink if one of its parents is accessed via symlink. This PR makes sure that we also evict content referenced like this.

Fixes https://github.com/gradle/gradle/issues/26201

## Implementation notes

Adding the additional check at the end of the build introduced a performance regression. To fix it, I moved the cleanup of the VFS (which includes more than just removing symlinked content) to happen between builds.

This means that the daemon will report the build to have been finished, and let the client proceed while the cleanup is running async. It also makes sure that the cleanup is finished by the time we start the next build.

In the typical case where the developer does some work between builds the work happening in the background should not impact the developer's work. The expected result is a small performance improvement. It's more noticeable for mostly up-to-date large builds, naturally.

### Update

Post-build cleanup is now split in two parts:

- Non symlink content is still evicted before the build is finished. This includes unsupported file systems, stuff that is not watched, and watched hierarchies over limit.
- Symlinked content (direct and indirect) is evicted after the build is finished.

Since very little of our tests rely on symlinks being evicted, not much needs to be updated. Specifically, tests relying on console logs will keep functioning.

The only somewhat meaningful difference is that the stats we print at the end of the build about what was retained in the VFS will still include any symlinked content. This will show up as removed between the builds when we print similar stats at the beginning of the build. Since nobody should rely on these stats, this should not be a big problem.

## TODO

- [x] Let continuous build run the cleanup between builds, too.
- [x] Do we need to trigger it in other places manually? TestKit? `ProjectBuilder`?
- [X] How do we track in-between-builds work via build operations? Probably we don't?
- [X] Figure out what to do with the stats we print in `WatchingVirtualFileSystem. beforeBuildFinished()`. They used to print stats about the post-cleanup state. Some options:
  - Remove these stats. We don't really use them much anyway.
  - Print them at the start of the build instead, when the cleanup has finished. They are not really interesting for the last build anyway. And printing it at the beginning would allow us to send the logs back to the client, too. (The logs from the in-between-build operations only appear in the daemon log now.)

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by lptr** [comment]
@bot-gradle test APT please.

**Comment by lptr** [comment]
@bot-gradle test APT please.

**Comment by lptr** [comment]
Looks like the simple approach leads to 5-10% regression in some important performance tests, since we are spending more time at the end of the build reading the file system:

image

Some things we could try:

1. parallelize the extra work
2. ~~keep track of the state of the file system for all the parents we look up~~ doing this would require watching those locations; something that would be a lot more expensive
3. move VFS eviction from "before build finished" to "between builds"

I think the we should attempt 3) next, as for real-world builds it would actually lead to slightly faster builds from the user's perspective.

**Comment by lptr** [comment]
@bot-gradle test QFL please.

**Comment by lptr** [comment]
@bot-gradle test QFL please.

File: platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/DaemonServices.java [hunk 1]
```diff
@@ -136,17 +137,32 @@ protected GarbageCollectorMonitoringStrategy createGarbageCollectorMonitoringStr
...
+            new HandleInvalidateVirtualFileSystem(gradleUserHomeScopeServiceRegistry),
             new HandleCancel(),
+            new EstablishBuildEnvironment(processEnvironment),
+            cleanUpVirtualFileSystemAfterBuild,
             new HandleReportStatus(),
             new ReturnResult(),
             new StartBuildOrRespondWithBusy(daemonDiagnostics), // from this point down, the daemon is 'busy'
-            new EstablishBuildEnvironment(processEnvironment),
             new LogToClient(loggingManager, daemonDiagnostics), // from this point down, logging is sent back to the client
             new LogAndCheckHealth(healthStats, healthCheck, runningStats),
```
Comment (wolfs) [inline platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/DaemonServices.java:167]: ❓ Why did you need to change the order here? What is the effect of changing the order?

File: platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/exec/CleanUpVirtualFileSystemAfterBuild.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        try {
+            execution.proceed();
+        } finally {
+            startAsyncCleanupAfterBuild(build);
+        }
+    }
+
+    private void startAsyncCleanupAfterBuild(Build build) {
+        userHomeServiceRegistry.getCurrentServices().ifPresent(serviceRegistry ->
+            pendingCleanup = CompletableFuture.runAsync(() -> {
```
Comment (wolfs) [inline platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/exec/CleanUpVirtualFileSystemAfterBuild.java:70]: 🤔 This looks somewhat hacky. Shouldn't we have already some infrastructure that allows us running stuff between builds? I am in particular worried here that we use the common pool instead of one of Gradle's executor pools to run the cleanup action. Moreover, the mutable state to hand the future from the last build to the current build seems strange.
Do we do anything like this somewhere else in Gradle?

File: platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/DaemonServices.java (hunk 1 above)
Comment (lptr) [reply platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/DaemonServices.java:167]: I needed this to get to the system property specifying the number of allowed watched hierarchies.

This should not be needed anymore, since I split up the cleanup between "before build finished" cleanup (all the existing cleanup, minus symlink cleanup) and "after build finished" cleanup (direct + indirect symlink cleanup).

Reverted.

File: platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/exec/CleanUpVirtualFileSystemAfterBuild.java (hunk 2 above)
Comment (lptr) [reply platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/exec/CleanUpVirtualFileSystemAfterBuild.java]: > 🤔 This looks somewhat hacky. Shouldn't we have already some infrastructure that allows us running stuff between builds?

I did some research and asked people about an "in-between builds" infrastructure, but nobody had an idea for it.


> I am in particular worried here that we use the common pool instead of one of Gradle's executor pools to run the cleanup action.

We are not using the common pool. Instead, we are creating a single-threaded managed pool, see the constructor.

> Moreover, the mutable state to hand the future from the last build to the current build seems strange.

I see what you mean. I tried to make it more obvious how the future is supposed to work, PTAL.

Note that this class is not thread safe, that's why it should be okay to handle the future in this way.

> Do we do anything like this somewhere else in Gradle?

I'm not aware of another example of doing something like this in Gradle.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test PT please.

**Comment by lptr** [comment]
@bot-gradle test QFL PT please.

**Review by ov7a — APPROVED** [review]

**Review by wolfs — APPROVED** [review]

**Comment by lptr** [comment]
@bot-gradle run APT please.

**Comment by lptr** [comment]
@bot-gradle test APT please.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/CrossVersionPerformanceTestRunner.groovy [hunk 3]
```diff
@@ -95,6 +97,16 @@ class CrossVersionPerformanceTestRunner extends PerformanceTestSpec {
...
+
+        // Pause between builds for the following reasons:
+        //
+        // - better simulate a normal development lifecycle where builds are rarely executed
+        //   back-to-back, as there is typically some work being done between the builds
+        // - give file system events a chance to arrive; otherwise we might detect them during
+        //   the build, and invalidating the VFS
+        // - let VFS cleanup (that happens between builds) finish
+        //
+        addBuildMutator { __ -> new DelayBeforeBuildMutator(500, MILLISECONDS) }
```
Comment (asodja) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/CrossVersionPerformanceTestRunner.groovy:109]: 🤔  Should we set that mutator directly in the `GradleBuildExperimentRunner`?
Otherwise other runners like `CrossBuildPerformanceTestRunner` doesnt' have this delay I believe

Comment (lptr) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/CrossVersionPerformanceTestRunner.groovy]: Good point, I updated the PR, PTAL.

**Comment by lptr** [comment]
@bot-gradle test APT please.

**Review by asodja — APPROVED** [review]
LGTM

**Comment by lptr** [comment]
@bot-gradle help

**Comment by lptr** [comment]
@bot-gradle test PT please.

**Comment by lptr** [comment]
@bot-gradle merge this please.

**Comment by lptr** [comment]
@bot-gradle merge this please.

**Comment by lptr** [comment]
@bot-gradle cancel merge.

**Comment by lptr** [comment]
@bot-gradle cancel

**Comment by lptr** [comment]
@bot-gradle merge please.
